vyperlang/vyper
# [WIP/POC] feat: implement offchain events

THIS IS A PROOF OF CONCEPT!

shadow events are defined like
```
shadow Foo:
    ...
```

and used just like regular events(!)

at codegen time, depending on the mode (`--shadow-mode`), they will either be stripped out (default) or converted to LOG instructions (`--shadow-mode`).

they could also be implemented syntactically using decorators but that resulted in more complicated changes to the tokenizer, this is easier as a POC.

the implementation depends on `--experimental-codegen` because the architecture is more pluggable there (an addition of just 20 loc pass!).

### What I did

### How I did it

### How to verify it
run with shadow mode and without, check opcodes output.
![carbon - 2023-12-08T200338 930](https://github.com/vyperlang/vyper/assets/3867501/51e120ac-10e7-4302-ba5f-f1e4a7355699)


### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Description for the changelog

### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->]()
